### ISSUE TITLE
flake8 --exclude and --per-file-ignores options are incompatible

### ISSUE THREAD

**Issue by felker** [flycheck/flycheck#1644]
## Bug description
I have been using a `setup.cfg` file at the root of my project directory successfully for a while now to specify the `ignore` option of `flake8`, which flycheck respects. 

In January of 2019, `flake8` 3.7.0 added the `--per-file-ignores` option. See discussion https://stackoverflow.com/questions/48153886/flake8-ignore-specific-warning-for-entire-file

I wanted to use this option with flycheck, but it does not appear to respect this setting, nor the older `--exclude` setting. 
 
## Steps to reproduce
Here is an example `setup.cfg`, with an aggressive `exclude` rule that should actually ignore all Python files:

```
[flake8]
max-line-length = 79
exclude = *.py
ignore =
       E121,
       E123,
       E126,
       E226,
       E241,
       E704,
       E731,
       W503
per-file-ignores =
       test.py:E402
```

Example `test.py`:
```python
import sys
```

Opening `test.py`,  the line is highlighted by Flycheck with:
> 'sys' imported but unused [F401]

`flycheck-compile` on the checker:
```
python -c import\ sys\;sys.path.pop\(0\)\;import\ runpy\;runpy.run_module\(\"flake8\"\) --format\=default --config /home/kfelker/plasma-python/setup.cfg - < \
/home/kfelker/plasma-python/test.py
stdin:1:1: F401 'sys' imported but unused
```

## Expected behavior
From the command line, `flake8 test.py` correctly returns no violations.

**Comment by felker** [comment]
Related #721, but I tried the workaround suggested in https://github.com/flycheck/flycheck/issues/721#issuecomment-129556743 and it did not work.

**Comment by fmdkdd** [comment]
Thanks for the report, can confirm.

The issue is that we pass the content of the buffer via stdin, so flake8 doesn't know the filename:

```
$ flake8 - < test.py
stdin:1:1: F401 'sys' imported but unused
```

Luckily, the flag `--stdin-display-name` solves the problem:

```
$ flake8 --stdin-display-name=test.py - < test.py
```

Fixed in #1645.

Sent

**Comment by felker** [comment]
Great! It may be worth documenting that this fix probably won't be compatible with the `flake8` options when they are set to relative paths, as opposed to simple file names.

**Comment by fmdkdd** [comment]
@felker Can you elaborate with an example?

**Comment by felker** [comment]
Sure, adapting an example from the [`flake8` documentation](http://flake8.pycqa.org/en/latest/user/options.html?highlight=per-file-ignores#cmdoption-flake8-per-file-ignores).

`project/setup.cfg`:
```
[flake8]
per-file-ignores =
     */__init__.py:F401
```

`project/test/__init__.py`:
```python
import sys
```

When opening `test/__init__.py` in Emacs, Flycheck returns:
> 'sys' imported but unused [F401] 

whereas the command line interface `flake8 .` returns no errors when run in the `project/` directory.  It does report the same diagnostic when executed from `project/test/`. I suppose Flycheck's behavior here is consistent with the statement from https://github.com/flycheck/flycheck/issues/721#issuecomment-129554169:
> Flycheck runs flake8 from the default directory of the buffer. Normally that's the directory the file is contained in.

**Comment by fmdkdd** [comment]
Thank you for providing an example.

Using the directory containing the flake8 configuration file as working directory fixes it.   Do you think we should go for that fix, or are there any adverse effects you can think of?

**Comment by felker** [comment]
I think this makes the most sense, as long as it is documented. I cannot think of any obvious drawbacks.

**Comment by fmdkdd** [comment]
Thanks, I've sent a patch you can test to make sure if you have time ;)

**Comment by cpitclaudel** [comment]
I think that's a Flake8 bug, FWIW.  Those paths in the configuration file should be relative to that configuration file.

**Comment by fmdkdd** [comment]
>  Those paths in the configuration file should be relative to that configuration file.

Yes, that is my expectation as well.  But yeah maybe it's better to report upstream than workaround it.
